vllm_ascend/torchair/torchair_model_runner.py

```diff
@@ -0,0 +1,340 @@
+# Adapted from vllm-project/vllm/vllm/worker/gpu_model_runner.py
+#
+
+import types
+
+import torch
+import torch._dynamo.cache_size
+import torch.nn as nn
+import torch_npu
+import vllm.envs as envs_vllm
+from vllm.config import VllmConfig
+from vllm.distributed import get_tensor_model_parallel_world_size
+from vllm.logger import logger
+
+import vllm_ascend.envs as envs_ascend
+from vllm_ascend.ascend_config import get_ascend_config
+from vllm_ascend.platform import NPUPlatform
+from vllm_ascend.torchair.utils import (check_torchair_cache_exist,
+                                        converting_weight_acl_format_310p,
+                                        write_kv_cache_bytes_to_file)
+from vllm_ascend.utils import (ACL_FORMAT_FRACTAL_ND, ACL_FORMAT_FRACTAL_NZ,
+                               is_310p)
+from vllm_ascend.worker.model_runner_v1 import NPUModelRunner
+
+if is_310p():
+    torch_npu.npu.set_compile_mode(jit_compile=False)
+    ACL_FORMAT = ACL_FORMAT_FRACTAL_NZ
+else:
+    ACL_FORMAT = ACL_FORMAT_FRACTAL_ND
+
+
+class NPUTorchairModelRunner(NPUModelRunner):
+
+    def __init__(self, vllm_config: VllmConfig, device: torch.device):
+        """Init torchair related parameters"""
+        super().__init__(vllm_config, device)
+
+        ascend_config = get_ascend_config()
+        self.new_kv_cache_bytes = -1
+        self.padded_batch_size = -1
+        self.torchair_compiled_model = None  # type: ignore
+        self.torchair_compiled_models = {}  # type: ignore
+        self.use_cached_npu_graph = ascend_config.torchair_graph_config.use_cached_graph
+        self.torchair_graph_batch_sizes = ascend_config.torchair_graph_config.graph_batch_sizes
+        if ascend_config.torchair_graph_config.graph_batch_sizes_init:
+            self._init_torchair_graph_batch_sizes()
+        if len(self.torchair_graph_batch_sizes) == 0:
+            # TODO(zzzzwwjj): check torchair_graph_batch_sizes init code
+            self.torchair_graph_batch_sizes = [self.max_num_reqs]
+        torch._dynamo.cache_size.config.cache_size_limit += len(
+            self.torchair_graph_batch_sizes)
+        torch._dynamo.config.capture_dynamic_output_shape_ops = True
+        torch._logging.set_logs(
+            recompiles=envs_ascend.VLLM_ASCEND_TRACE_RECOMPILES)
+
+    def _generate_extra_builder_kwargs(self, total_num_scheduled_tokens,
+                                       with_prefill):
+        """Generate graph_pad_size for torchair case."""
+        extra_builder_kwargs = {}
+
+        if self.dp_size > 1:
+            max_num_tokens, with_prefill = self._get_forward_metadata_across_dp(
+                total_num_scheduled_tokens, with_prefill)
+            extra_builder_kwargs['max_num_tokens_across_dp'] = max_num_tokens
+            extra_builder_kwargs['with_prefill_across_dp'] = with_prefill
+        if not with_prefill:
+            if self.dp_size > 1:
+                self.padded_batch_size = self._select_torchair_padded_batch_size(
```

Comment:
qq: It seems we change padded_batch_size to self.padded_batch_size here, should we do some re-init actions?